{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IN4030 - Partial exam 1  \n",
    "Markus Sverdvik Heiervang - markuhei\n",
    "***  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import numpy as np\n",
    "from typing import List, Tuple\n",
    "from collections import namedtuple"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 1: Database searches (8p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a) \n",
    "Retrieve the protein sequence with accession number NP_001264426 from the National Center for\n",
    "Biotechnology Information (NCBI) using their website at https://www.ncbi.nlm.nih.gov/ and its search\n",
    "functions. The sequence should be 247 amino acids long. Save it in a FASTA-formatted text file. Show the\n",
    "formatted sequence in the report. (1p)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'>NP_001264426.1 alpha-ketoglutarate-dependent dioxygenase alkB homolog 2 isoform 1 [Gallus gallus]\\r\\nMDRFVVKRSAEEPGGDGKKPRLEEEAGGLPHPSQPSQEIRAQGLSLEYRLLFGRAEADAIFQQLEKEVEY\\r\\nFEGEQTKLHVFGKWHNIPRKQVTYGDPELTYTYSGVTFSPKPWIPVLNHIRDRLVLETGHTFNFVLINRY\\r\\nKDGEDHIGEHRDDEKELVPRSPIASVSFGACRDFVFRHCDSRGKNATRHIKPIRLQLAHGSLLMMKYPTN\\r\\nVYWYHSLPIRRRVLAPRINLTFRKMMAVDKKDSILTT\\r\\n\\r\\n'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "link = \"https://www.ncbi.nlm.nih.gov/search/api/download-sequence/?db=protein&id=NP_001264426.1\"\n",
    "response = requests.get(link)\n",
    "\n",
    "# The content is in binary format so we decode it with utf8\n",
    "content_str = response.content.decode(\"UTF8\")\n",
    "\n",
    "# The data we got is already in fasta format, so we write it direclty to a fasta file\n",
    "with open(\"NP_001264426.fasta\", \"w+\") as f:\n",
    "    f.write(content_str)\n",
    "\n",
    "content_str"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b) \n",
    "Which variant of BLAST should be used to search the UniProKB/Swiss-Prot database with the retrieved\n",
    "sequence as the query? (1p)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(TODO)  \n",
    "\n",
    "blastp, sice it its designed for proteins as opposed to blastn that's designed for nucleotides"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c) \n",
    "Use BLAST at https://blast.ncbi.nlm.nih.gov/Blast.cgi to search the UniProKB/Swiss-Prot database with the\n",
    "retrieved sequence as the query. Algorithm parameters must be adjusted as follows: \n",
    "```\n",
    "“Expect threshold” = 1,\n",
    "\"Word size\" = 3, \n",
    "\"Compositional adjustments\" = \"No adjustment\",\n",
    "“Low complexity regions” = OFF.\n",
    "scoring matrix = BLOSUM62, \n",
    "gap costs: Existence=11 Extension=1 \n",
    "```\n",
    "How many\n",
    "hits do you get? (1p)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A: I got 10 results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from Bio.Blast.NCBIWWW import qblast\n",
    "# result = qblast(\n",
    "#     program=\"blastp\",\n",
    "#     database=\"UniProKB/Swiss-Prot\",\n",
    "#     sequence=content_str,\n",
    "    \n",
    "#     expect=1,\n",
    "#     word_size=3,\n",
    "#     matrix_name=\"BLOSUM62\",\n",
    "#     gapcosts=\"11 1\",\n",
    "# )\n",
    "# result\n",
    "\n",
    "# results = list(result)\n",
    "# results\n",
    "\n",
    "# help(qblast)\n",
    "\n",
    "# from IPython.core.display import HTML, display\n",
    "# display(HTML(\"\".join(results)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### d) \n",
    "What is the accession number of the best match with a human sequence? Save the sequence in a file. (1p)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q6NS38.1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### e) \n",
    "\n",
    "Show the BLAST alignment between the query protein and the best matching human protein. (1p)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "Alignment statistics for match #1\n",
    "Score\tExpect\tIdentities\tPositives\tGaps\n",
    "319 bits(817)\t1e-109\t163/262(62%)\t192/262(73%)\t23/262(8%)\n",
    "Query  1    MDRFVVKRS--------------AEEP---GGDG----KKPRLEEEAGGLPHPSQPS-QE  38\n",
    "            MDRF+VK +               EEP   GGD     K+PR E    G  H + PS + \n",
    "Sbjct  1    MDRFLVKGAQGGLLRKQEEQEPTGEEPAVLGGDKESTRKRPRREAPGNG-GHSAGPSWRH  59\n",
    "\n",
    "Query  39   IRAQGLSLEYRLLFGRAEADAIFQQLEKEVEYFEGEQTKLHVFGKWHNIPRKQVTYGDPE  98\n",
    "            IRA+GL   Y +LFG+AEAD IFQ+LEKEVEYF G   ++ VFGKWH++PRKQ TYGD  \n",
    "Sbjct  60   IRAEGLDCSYTVLFGKAEADEIFQELEKEVEYFTGALARVQVFGKWHSVPRKQATYGDAG  119\n",
    "\n",
    "Query  99   LTYTYSGVTFSPKPWIPVLNHIRDRLVLETGHTFNFVLINRYKDGEDHIGEHRDDEKELV  158\n",
    "            LTYT+SG+T SPKPWIPVL  IRD +   TG TFNFVLINRYKDG DHIGEHRDDE+EL \n",
    "Sbjct  120  LTYTFSGLTLSPKPWIPVLERIRDHVSGVTGQTFNFVLINRYKDGCDHIGEHRDDERELA  179\n",
    "\n",
    "Query  159  PRSPIASVSFGACRDFVFRHCDSRGKNATRHIKPIRLQLAHGSLLMMKYPTNVYWYHSLP  218\n",
    "            P SPIASVSFGACRDFVFRH DSRGK+ +R +  +RL LAHGSLLMM +PTN +WYHSLP\n",
    "Sbjct  180  PGSPIASVSFGACRDFVFRHKDSRGKSPSRRVAVVRLPLAHGSLLMMNHPTNTHWYHSLP  239\n",
    "\n",
    "Query  219  IRRRVLAPRINLTFRKMMAVDK  240\n",
    "            +R++VLAPR+NLTFRK++   K\n",
    "Sbjct  240  VRKKVLAPRVNLTFRKILLTKK  261\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### f)  \n",
    "What is the score (raw alignment score), the bitscore and the E-value of this alignment? (3p)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "319, 817, E-value: 1e-109"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 2: Alignments (22p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Write a program to compute the optimal local alignment of two amino acid sequences using the Smith-Waterman algorithm with affine gap penalties. This issue can be solved in at least two different ways. The program shall read\n",
    "the two sequences in FASTA format from two separate files and print the optimal alignment score and the\n",
    "actual alignment. You only need to show one alignment if there are multiple optimal alignments with the same\n",
    "score. The alignment should be presented in a comprehensible manner, preferably in a way similar to BLAST.\n",
    "The program shall use the BLOSUM62 scoring matrix and an affine gap penalty function with gap opening\n",
    "penalty of 11 and a gap extension penalty of 1, i.e. gap penalty = 11 + gap length (12p)  \n",
    "\n",
    "* You are not allowed to use libraries or other imported code for reading the FASTA files or performing the actual alignment. \n",
    "\n",
    "* The BLOSUM62 matrix can be downloaded from the NCBI at https://www.ncbi.nlm.nih.gov/Class/FieldGuide/BLOSUM62.txt  \n",
    "\n",
    "* Ideally, the scoring matrix is read from a specified file and the gap penalties are specified as options to the program. Information about testing your program is provided on page 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See\n",
    "chapter 5 in the book for details (especially lesson 12.5 “Penalizing Insertions and Deletions in Sequence\n",
    "Alignments” at https://www.bioinformaticsalgorithms.org/bioinformatics-chapter-5)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>R</th>\n",
       "      <th>N</th>\n",
       "      <th>D</th>\n",
       "      <th>C</th>\n",
       "      <th>Q</th>\n",
       "      <th>E</th>\n",
       "      <th>G</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>...</th>\n",
       "      <th>P</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>V</th>\n",
       "      <th>B</th>\n",
       "      <th>Z</th>\n",
       "      <th>X</th>\n",
       "      <th>*</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>-1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-4</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>9</td>\n",
       "      <td>-3</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>6</td>\n",
       "      <td>-2</td>\n",
       "      <td>-4</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>8</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>2</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-4</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>-1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-4</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>2</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>-3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3</td>\n",
       "      <td>-3</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z</th>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-3</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-3</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>...</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>-4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   A  R  N  D  C  Q  E  G  H  I  ...  P  S  T   W  Y  V  B  Z  X  *\n",
       "A  4 -1 -2 -2  0 -1 -1  0 -2 -1  ... -1  1  0  -3 -2  0 -2 -1  0 -4\n",
       "R -1  5  0 -2 -3  1  0 -2  0 -3  ... -2 -1 -1  -3 -2 -3 -1  0 -1 -4\n",
       "N -2  0  6  1 -3  0  0  0  1 -3  ... -2  1  0  -4 -2 -3  3  0 -1 -4\n",
       "D -2 -2  1  6 -3  0  2 -1 -1 -3  ... -1  0 -1  -4 -3 -3  4  1 -1 -4\n",
       "C  0 -3 -3 -3  9 -3 -4 -3 -3 -1  ... -3 -1 -1  -2 -2 -1 -3 -3 -2 -4\n",
       "Q -1  1  0  0 -3  5  2 -2  0 -3  ... -1  0 -1  -2 -1 -2  0  3 -1 -4\n",
       "E -1  0  0  2 -4  2  5 -2  0 -3  ... -1  0 -1  -3 -2 -2  1  4 -1 -4\n",
       "G  0 -2  0 -1 -3 -2 -2  6 -2 -4  ... -2  0 -2  -2 -3 -3 -1 -2 -1 -4\n",
       "H -2  0  1 -1 -3  0  0 -2  8 -3  ... -2 -1 -2  -2  2 -3  0  0 -1 -4\n",
       "I -1 -3 -3 -3 -1 -3 -3 -4 -3  4  ... -3 -2 -1  -3 -1  3 -3 -3 -1 -4\n",
       "L -1 -2 -3 -4 -1 -2 -3 -4 -3  2  ... -3 -2 -1  -2 -1  1 -4 -3 -1 -4\n",
       "K -1  2  0 -1 -3  1  1 -2 -1 -3  ... -1  0 -1  -3 -2 -2  0  1 -1 -4\n",
       "M -1 -1 -2 -3 -1  0 -2 -3 -2  1  ... -2 -1 -1  -1 -1  1 -3 -1 -1 -4\n",
       "F -2 -3 -3 -3 -2 -3 -3 -3 -1  0  ... -4 -2 -2   1  3 -1 -3 -3 -1 -4\n",
       "P -1 -2 -2 -1 -3 -1 -1 -2 -2 -3  ...  7 -1 -1  -4 -3 -2 -2 -1 -2 -4\n",
       "S  1 -1  1  0 -1  0  0  0 -1 -2  ... -1  4  1  -3 -2 -2  0  0  0 -4\n",
       "T  0 -1  0 -1 -1 -1 -1 -2 -2 -1  ... -1  1  5  -2 -2  0 -1 -1  0 -4\n",
       "W -3 -3 -4 -4 -2 -2 -3 -2 -2 -3  ... -4 -3 -2  11  2 -3 -4 -3 -2 -4\n",
       "Y -2 -2 -2 -3 -2 -1 -2 -3  2 -1  ... -3 -2 -2   2  7 -1 -3 -2 -1 -4\n",
       "V  0 -3 -3 -3 -1 -2 -2 -3 -3  3  ... -2 -2  0  -3 -1  4 -3 -2 -1 -4\n",
       "B -2 -1  3  4 -3  0  1 -1  0 -3  ... -2  0 -1  -4 -3 -3  4  1 -1 -4\n",
       "Z -1  0  0  1 -3  3  4 -2  0 -3  ... -1  0 -1  -3 -2 -2  1  4 -1 -4\n",
       "X  0 -1 -1 -1 -2 -1 -1 -1 -1 -1  ... -2  0  0  -2 -1 -1 -1 -1 -1 -4\n",
       "* -4 -4 -4 -4 -4 -4 -4 -4 -4 -4  ... -4 -4 -4  -4 -4 -4 -4 -4 -4  1\n",
       "\n",
       "[24 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "blosum62 = pd.read_csv(\"BLOSUM62.txt\", delim_whitespace=True, comment=\"#\")\n",
    "blosum62"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 2)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def argmax(matrix: List[List[float]]) -> Tuple[int, int]:\n",
    "    \"Exhaustively searches for the highest number in a matrix and outputs it's indices\"\n",
    "    m = len(matrix)\n",
    "    n = len(matrix[0])\n",
    "    return max(((i, j) for i in range(m) for j in range (n)), key=lambda t: matrix[t[0]][t[1]])\n",
    "\n",
    "argmax([[1,2,3]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Look at https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm\n",
    "def optimal_alignment(seq1: str, seq2: str, \n",
    "                      substitution_matrix: pd.DataFrame = blosum62, \n",
    "                      extension_penalty: int = 1,\n",
    "                      opening_penalty: int = 11) -> float:\n",
    "    \"\"\"\n",
    "    Computes the optimal alignment between two nucleotide/amino acid sequences\n",
    "    using the BLOSUM62 scoring matrix\n",
    "    \n",
    "    It uses an affine gap penalty function with gap opening penalty of 11 and gap\n",
    "    extension penalty of 1.\n",
    "    \"\"\"\n",
    "    \n",
    "    m = len(seq1)\n",
    "    n = len(seq2)\n",
    "    matrix = [[0] * n for i in range(m)]\n",
    "    \n",
    "    \n",
    "    def affine_gap_penalty(n): \n",
    "        return opening_penalty + n * extension_penalty\n",
    "\n",
    "    # TODO: REPLACE WITH ENUMERATE\n",
    "\n",
    "    \n",
    "    for i, a1 in enumerate(seq1[1:]):\n",
    "        for j, a2 in enumerate(seq2[1:]):\n",
    "            #amino acid 1 and 2\n",
    "            match_score = substitution_matrix[a1][a2]\n",
    "            \n",
    "            penalty = extension_penalty\n",
    "            \n",
    "            diag = matrix[i][j] + match_score\n",
    "            top = max(range(j+1), key=lambda k: matrix[i][k] - affine_gap_penalty(j-k))\n",
    "            left = max(range(i+1), key=lambda k: matrix[k][j] - affine_gap_penalty(i-k))\n",
    "            matrix[i+1][j+1] = max(0, diag, top, left)\n",
    "    \n",
    "    i, j = argmax(matrix)\n",
    "    max_score = matrix[i][j]\n",
    "    \n",
    "    # Traceback\n",
    "    align1, align2 = [], []\n",
    "    \n",
    "    current_score = max_score\n",
    "\n",
    "    while current_score > 0:\n",
    "        directions = {\n",
    "            \"UP\": matrix[i-1][j],\n",
    "            \"LEFT\": matrix[i][j-1],\n",
    "            \"DIAG\": matrix[i-1][j-1]\n",
    "        }\n",
    "\n",
    "        direction, current_score = max(directions.items(), key=lambda x: x[1])\n",
    "\n",
    "        if direction == \"UP\":\n",
    "            i -= 1\n",
    "            align1.append(seq1[i])\n",
    "            align2.append(\"-\")\n",
    "        elif direction == \"LEFT\":\n",
    "            j -= 1\n",
    "            align1.append(\"-\")\n",
    "            align2.append(seq2[j])            \n",
    "        elif direction == \"DIAG\":\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "            align1.append(seq1[i])\n",
    "            align2.append(seq2[j])\n",
    "        \n",
    "    align1 = \"\".join(reversed(align1))\n",
    "    align2 = \"\".join(reversed(align2))\n",
    "    \n",
    "    return align1, align2, max_score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Run the program on the querysequence and the human sequence from exercise1.What score do you get? Show the alignment. (2p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('M--------------------------------DRFVVKRSAE-EP-GGDGK-KPRLEEEAG-GLPHPSQPSQE-IRAQGLSLEYRLLFGRAEADAIFQQLEKEVEYFEGEQTKLHVFGKWHNIPRKQVTYGDPELTYTYSGVTFSPKPWIPVLNHIRDRLVLETGHTFNFVLINRYKDGEDHIGEHRDDEKELVPRSPIASVSFGACRDFVFRHCDSRGKNATRHIKPIRLQLAHGSLLMMKYPTNVYWYHSLPIRRRVLAPRINLTFRKMMAVD',\n",
       " '-DRFLVKGAQGGLLRKQEEQEPTGEEPAVLGGD-----K----ESTR-----KRPRREAPGNGG--HSAGPSWRHIRAEGLDCSYTVLFGKAEADEIFQELEKEVEYFTGALARVQVFGKWHSVPRKQATYGDAGLTYTFSGLTLSPKPWIPVLERIRDHVSGVTGQTFNFVLINRYKDGCDHIGEHRDDERELAPGSPIASVSFGACRDFVFRHKDSRGKSPSRRVAVVRLPLAHGSLLMMNHPTNTHWYHSLPVRKKVLAPRVNLTFRKILLTK',\n",
       " 845)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FastaSequence = namedtuple(\"FastaSequence\", [\"meta\", \"data\"])\n",
    "\n",
    "def read_fasta(filename: str):\n",
    "    with open(filename, \"r\", encoding=\"UTF8\") as f:\n",
    "        data = f.read().split(\"\\n\")\n",
    "    return FastaSequence(data[0], \"\".join(data[1:]))\n",
    "\n",
    "seq1 = read_fasta(\"original_sequence.fasta\").data\n",
    "seq2 = read_fasta(\"best_human_sequence.fasta\").data\n",
    "\n",
    "optimal_alignment(seq1, seq2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "c) Compare thescore you got with the rawscore from BLAST in exercise1 and comment on any discrepancies. (2p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "d) Compare the actual alignmentyou got with the alignment from BLAST and comment on any discrepancies.(2p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "e)Retrieve the ALKB_ECOLI (P05050) sequence from Escherichia coliin FASTA format. Use your program to compare this sequence with the sequence fromexercise1a. What score do you get? Show the alignment. Why did the sequence from E. coli not showup in the search in exercise 1? (4p)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
